scipy/stats/_continuous_distns.py

```diff
@@ -7964,6 +7965,136 @@ def _sf(self, x, chi):
+    def _rvs_scalar_small_chi(self, chi, numsamples=None):
+        # use Vaduva's rejection method, see Devroye:
+        # Non-Uniform Random Variate Generation, 1986, section II.3.2.
+        # write: self.pdf = c * g(x) * (1 - H(x)), where
+        # H is a distribution function and g is a density
+        # g(x) = d1 * chi**2 * x * exp(-chi**2 * (1 - x**2) / 2), 0 <= x <= 1
+        # H(x) = 1 - sqrt(1 - x**2), 0 <= x <= 1
+        # Integrating g, we get:
+        # G(x) = d1 * exp(-chi**2 * (1 - x**2) / 2) - d2
+        # d1 and d2 are determined by G(0) = 0 and G(1) = 1
+        # d1 = 1 / (1 - exp(-0.5 * chi**2))
+        # d2 = 1 / (exp(0.5 * chi**2) - 1)
+        # => G(x) = (exp(chi**2 * x**2 /2) - 1) / (exp(chi**2 / 2) - 1)
+        # expected number of iterations is c
+        # c = -np.expm1(-0.5 * chi**2) * chi / (_norm_pdf_C * _argus_phi(chi))
+        # note that both G and H can be inverted easily, so we can sample
+        # rvs from these distributions
+        # H_inv(y) = sqrt(1 - (1 - y)**2)
+        # G_inv(y) = sqrt(2 * log(1 + A * y) / chi**2), A = exp(c**2 / 2) - 1
+
+        A = np.expm1(chi**2 / 2)
+        if not np.isfinite(A):
+            raise ValueError("Overflow occured for chi={}. "
+                             "Sampling does not work for such large "
+                             "values of chi".format(chi))
+        size1d = tuple(np.atleast_1d(numsamples))
+        N = int(np.prod(size1d))
+        x = np.zeros(N)
+        simulated = 0
+        # apply Vaduva's rejection method
+        while simulated < N:
+            k = N - simulated
+            u = self._random_state.random_sample(size=k)
+            v = self._random_state.random_sample(size=k)
+            # generate rvs with density g by inverting G
+            z = 2 * np.log1p(v * A) / chi**2
+            rvs = np.sqrt(z)
+            # acceptance condition: rvs <= Y where Y has cdf H which gives
+            # accept = (rvs <= np.sqrt(1 - (1-u)**2))
+            # note that 1 - u is uniform as well, so this simplifies to:
+            accept = (z + u**2 <= 1)
+            num_accept = np.sum(accept)
+            if num_accept > 0:
+                x[simulated:(simulated + num_accept)] = rvs[accept]
+                simulated += num_accept
+
+        if self._size == ():
+            # return scalar if rvs() is called without size specified
+            return x[0]
+        return np.reshape(x, size1d)
+
+    def _rvs_scalar_large_chi(self, chi, numsamples=None):
+        # we build on the rejection algorithm in _rvs_scalar_small_chi
+        # G(x) = (exp(chi**2 * x**2 /2) - 1) / (exp(chi**2 / 2) - 1)
+        # H(x) = 1 - sqrt(1 - x**2), 0 <= x <= 1
+        # Note that G >= G* with
+        # G*(x) = (exp(chi**2 * x**2 /2) - 1) / exp(chi**2 / 2)
+        # Then G_inv <= G*_inv where the inverse of G* is given by
+        # G*_inv(y) = sqrt(2 * log(1 + exp(c**2 / 2) * y) / chi**2)
+        #           = sqrt(1 + 2 * log(y + exp(-c**2 / 2)))
```

Comment:
Is the `c` here the same as the `c` in  the description of `_rvs_scalar_small_chi`?
Or should it be `chi`?